File: c4-3-dandc.js
function findMax(arr, max) {
  if (!max) {
    max = 0;
  }
  if (arr.length === 1) {
    if (arr[0] > max) {
      return arr[0];
    }
    return max;
  } else {
    if (arr[0] > max) {
      max = arr[0];
    }
    return findMax(arr.slice(1), max);
  }
}

console.log(findMax([4, 1, 3, 5, 2]));

File: python/c4-3-dandc.py
def find_max(arr):
    def find_max_inner(arr, max_num):
        if len(arr) == 1:
            if arr[0] > max_num:
                return arr[0]
            else:
                return max_num
        else:
            new_max = max_num
            if arr[0] > max_num:
                new_max = arr[0]
            return find_max_inner(arr[1:], new_max)

    return find_max_inner(arr, 0)


print(find_max([4, 1, 3, 5, 2]))
